staging/src/k8s.io/apiserver/pkg/endpoints/discovery/aggregated/handler.go

```diff
@@ -0,0 +1,267 @@
+		rdm.apiGroups = make(map[string]*apidiscoveryv2beta1.APIGroupDiscovery)
+	}
+
+	if existing, groupExists := rdm.apiGroups[groupName]; groupExists {
+		// If this version already exists, replace it
+		versionExists := false
+
+		// Not very efficient, but in practice there are generally not many versions
+		for i := range existing.Versions {
+			if existing.Versions[i].Version == value.Version {
+				existing.Versions[i] = value
+				versionExists = true
+				break
+			}
+		}
+
+		if !versionExists {
+			existing.Versions = append(existing.Versions, value)
+		}
+
+	} else {
+		group := &apidiscoveryv2beta1.APIGroupDiscovery{
+			TypeMeta: metav1.TypeMeta{
+				Kind:       "APIGroupDiscovery",
+				APIVersion: "apidiscovery.k8s.io/v2beta1",
+			},
+			ObjectMeta: metav1.ObjectMeta{
+				Name: groupName,
+			},
+			Versions: []apidiscoveryv2beta1.APIVersionDiscovery{value},
+		}
+		rdm.apiGroups[groupName] = group
+		rdm.apiGroupNames = append(rdm.apiGroupNames, group.Name)
+
+	}
+
+	// Reset response document so it is recreated lazily
+	rdm.cachedGroupList = cachedGroupList{}
+}
+
+func (rdm *resourceDiscoveryManager) RemoveGroupVersion(apiGroup metav1.GroupVersion) {
+	rdm.lock.Lock()
+	defer rdm.lock.Unlock()
+
+	group, exists := rdm.apiGroups[apiGroup.Group]
+	if !exists {
+		return
+	}
+
+	for i := range group.Versions {
+		if group.Versions[i].Version == apiGroup.Version {
+			group.Versions = append(group.Versions[:i], group.Versions[i+1:]...)
+			break
+		}
+	}
+
+	if len(group.Versions) == 0 {
+		for i, g := range rdm.apiGroupNames {
+			if g == group.Name {
+				rdm.apiGroupNames = append(rdm.apiGroupNames[:i], rdm.apiGroupNames[i+1:]...)
+			}
+		}
+		delete(rdm.apiGroups, group.Name)
+	}
+
+	// Reset response document so it is recreated lazily
+	rdm.cachedGroupList = cachedGroupList{}
+}
+
+func (rdm *resourceDiscoveryManager) RemoveGroup(groupName string) {
+	rdm.lock.Lock()
+	defer rdm.lock.Unlock()
+
+	delete(rdm.apiGroups, groupName)
+
+	// Reset response document so it is recreated lazily
+	rdm.cachedGroupList = cachedGroupList{}
+}
+
+func (rdm *resourceDiscoveryManager) ServeHTTP(resp http.ResponseWriter, req *http.Request) {
+	rdm.lock.RLock()
+
+	// The cachedResponse is wiped out every time there might be a change to its
+	// contents.
+	if rdm.groups != nil {
+		defer rdm.lock.RUnlock()
+	} else {
+		rdm.lock.RUnlock()
+		func() {
+			// Document does not exist, recreate it
+			rdm.lock.Lock()
+			defer rdm.lock.Unlock()
+
+			// Check if another thread has populated the cache when we obtain the lock
+			if rdm.groups == nil {
+				// Re-order the apiGroups by their insertion order
+				groups := []apidiscoveryv2beta1.APIGroupDiscovery{}
+				for _, v := range rdm.apiGroupNames {
+					groups = append(groups, *rdm.apiGroups[v].DeepCopy())
+				}
+				rdm.groups = groups
+			}
+		}()
+		rdm.lock.RLock()
+		defer rdm.lock.RUnlock()
+	}
+
+	response := &apidiscoveryv2beta1.APIGroupDiscoveryList{
+		TypeMeta: metav1.TypeMeta{
```

Comment:
@smarterclayton I'm a bit confused whether we need to specify TypeMeta here. AFAIK, `responsewriters.WriteObjectNegotiated` is able to populate this field through the encoding procedure. I'm curious whether filling this information is a shortcut or is plain wrong.

Reply:
You should not have to specify TypeMeta - but it's good in your tests to verify the output to check that TypeMeta is correctly set on the response object and matches the expected transformation, e.g. as https://github.com/kubernetes/kubernetes/blob/1f74daccbcd317c9cf779a842e0c71b93414b60f/staging/src/k8s.io/apiserver/pkg/endpoints/apiserver_test.go#L1802